pkg/controller/keyvault/secret/keyvault_secret.go

```diff
@@ -0,0 +1,192 @@
+	errUpdateFailed  = "cannot update the Key Vault Secret"
+	errDeleteFailed  = "cannot delete the Key Vault Secret"
+)
+
+// SetupSecret adds a controller that reconciles KeyVaultSecret resources.
+func SetupSecret(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter, poll time.Duration) error {
+	name := managed.ControllerName(v1alpha1.KeyVaultSecretGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&v1alpha1.KeyVaultSecret{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.KeyVaultSecretGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient()}),
+			managed.WithPollInterval(poll),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	kube client.Client
+}
+
+func (c connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	_, auth, err := azure.GetAuthInfo(ctx, c.kube, mg)
+	if err != nil {
+		return nil, errors.Wrap(err, errConnectFailed)
+	}
+	cl := keyvault.New()
+	cl.Authorizer = auth
+	return &external{kube: c.kube, client: cl}, nil
+}
+
+type external struct {
+	kube   client.Client
+	client keyvaultapi.BaseClientAPI
+}
+
+func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.KeyVaultSecret)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotSecret)
+	}
+
+	secret, err := c.client.GetSecret(ctx, cr.Spec.ForProvider.VaultBaseURL, cr.Spec.ForProvider.Name, "" /* latest */)
+	if err != nil {
+		return managed.ExternalObservation{ResourceExists: false}, errors.Wrap(resource.Ignore(azure.IsNotFound, err), errGetFailed)
+	}
+
+	lateInit := false
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	secretclients.LateInitialize(&cr.Spec.ForProvider, secret)
+	if !cmp.Equal(currentSpec, &cr.Spec.ForProvider) {
+		if err := c.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, errUpdateSecretCRFailed)
+		}
+		lateInit = true
+	}
+
+	cr.Status.SetConditions(xpv1.Available())
+	cr.Status.AtProvider = secretclients.GenerateObservation(secret)
+
+	isUpToDate, err := secretclients.IsUpToDate(ctx, c.kube, cr.Spec.ForProvider, &secret)
+
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errCheckUpToDate)
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        isUpToDate,
+		ResourceLateInitialized: lateInit,
+	}, nil
+}
+
+func (c *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.KeyVaultSecret)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotSecret)
+	}
+
+	cr.Status.SetConditions(xpv1.Creating())
```

Comment:
crossplane-runtime can now [handle setting "creating" and "deleting" conditions](https://github.com/crossplane/crossplane-runtime/pull/242).